net/tor/src/opnsense/mvc/app/views/OPNsense/Tor/general.volt

```diff
@@ -0,0 +1,197 @@
+        formatTokenizersUI();
+        $('select.dropdownstyle').selectpicker('refresh');
+    });
+    ajaxCall(url="/api/tor/service/status", sendData={}, callback=function(data,status) {
+        updateServiceStatusUI(data['result']);
+    });
+
+    // link save button to API set action
+    [
+      {'selector': '#generalsaveAct', 'endpoint': '/api/tor/general/set', 'formid': 'general'},
+      {'selector': '#relaysaveAct', 'endpoint': '/api/tor/relay/set', 'formid': 'relay'}
+    ].forEach(function (cfg) {
+        $(cfg.selector).click(function(){
+            saveFormToEndpoint(url=cfg.endpoint, formid=cfg.formid,callback_ok=function(){
+                $(cfg.selector + " .saveAct_progress").addClass("fa fa-spinner fa-pulse");
+                ajaxCall(url="/api/tor/service/restart", sendData={}, callback=function(data,status) {
+                    ajaxCall(url="/api/tor/service/status", sendData={}, callback=function(data,status) {
+                        updateServiceStatusUI(data['result']);
+                    });
+                    $(cfg.selector + " .saveAct_progress").removeClass("fa fa-spinner fa-pulse");
+                });
+            });
+        });
+    });
+    $("#grid-hidden").UIBootgrid(
+        { 'search':'/api/tor/hiddenservice/searchservice',
+          'get':'/api/tor/hiddenservice/getservice/',
+          'set':'/api/tor/hiddenservice/setservice/',
+          'add':'/api/tor/hiddenservice/addservice/',
+          'del':'/api/tor/hiddenservice/delservice/',
+          'toggle':'/api/tor/hiddenservice/toggleservice/',
+          'options':{selection:false, multiSelect:false}
+        }
+    );
+    $("#grid-hiddenacl").UIBootgrid(
+        { 'search':'/api/tor/hiddenserviceacl/searchacl',
+          'get':'/api/tor/hiddenserviceacl/getacl/',
+          'set':'/api/tor/hiddenserviceacl/setacl/',
+          'add':'/api/tor/hiddenserviceacl/addacl/',
+          'del':'/api/tor/hiddenserviceacl/delacl/',
+          'toggle':'/api/tor/hiddenserviceacl/toggleacl/',
+          'options':{selection:false, multiSelect:false}
+        }
+    );
+    $("#grid-toracl").UIBootgrid(
+        { 'search':'/api/tor/socksacl/searchacl',
+          'get':'/api/tor/socksacl/getacl/',
+          'set':'/api/tor/socksacl/setacl/',
+          'add':'/api/tor/socksacl/addacl/',
+          'del':'/api/tor/socksacl/delacl/',
+          'toggle':'/api/tor/socksacl/toggleacl/',
+          'options':{selection:false, multiSelect:false}
+        }
+    );
+});
+
+</script>
+<ul class="nav nav-tabs" data-tabs="tabs" id="maintabs">
+    <li class="active"><a data-toggle="tab" href="#general">{{ lang._('General') }}</a></li>
+    <li><a data-toggle="tab" href="#acl">{{ lang._('ACL (SOCKS Proxy)') }}</a></li>
+    <li><a data-toggle="tab" href="#hidden">{{ lang._('Hidden Services') }}</a></li>
+    <li><a data-toggle="tab" href="#hiddenrouting">{{ lang._('Routing (Hidden Services)') }}</a></li>
```

Comment:
This looks weird, basically anything that needs explaining in "()" is not clear UX. maybe split pages / tabs or use the sub tabs that the proxy uses